### DISCUSSION TITLE
Detect thin rects as lines

### DISCUSSION THREAD

**Discussion by bellma-lilly** [jsvine/pdfplumber#839] [Ask for help with specific PDFs]
[redacted.pdf](https://github.com/jsvine/pdfplumber/files/10968650/redacted.pdf)
In the attached pdf, the lines in the table are created with thin rectangles rather than lines. I'm not sure what software was used to create this table.
```
table_settings={"vertical_strategy": "lines_strict", "horizontal_strategy": "lines_strict"}
for page in pdf.pages:
    tables = page.extract_tables(table_settings=table_settings)
```
With this code, no tables are found.

I have code ready for a PR which detects thin rectangles and uses them as lines (utils/geometry.py:
```
def test(e: T_obj) -> bool:
        dim = "height" if e["orientation"] == "v" else "width"
        et_correct = e["object_type"] == edge_type if edge_type is not None else True
        if(edge_type == "line" and e["object_type"] == 'rect_edge'):
            #maybe this rect is meant to be a line if the width is small
            if(max([p[0] for p in e['pts']]) - min([p[0] for p in e['pts']]) < x_tolerance or
               max([p[1] for p in e['pts']]) - min([p[1] for p in e['pts']]) < y_tolerance):
                et_correct = True
               ...
```

This change captures the table.
Shall l submit this PR?

**Comment by jsvine** [comment]
Hi @bellma-lilly, and thanks for your interest in this library. That is, in fact, the purpose of the `lines_strict` setting: To use *only* lines defined as lines and things that look like lines but are really rectangles. If you want to incorporate the rectangles, you can just omit those parts of the settings, sticking with the default approach. Doing so does, indeed, appear to detect the tables (although you might want to opt for further refinement, given some of the other quirks of the PDF):

```python
import pdfplumber
pdf = pdfplumber.open("redacted.pdf")
page = pdf.pages[0]
im = page.to_image()
im.debug_tablefinder()
```

![image](https://user-images.githubusercontent.com/534702/225005009-35e0a4fc-ccaf-4731-acb7-f0e1965f64da.png)

**Reply by bellma-lilly** [reply]
@jsvine I have been using pdfplumber for quite a while, and I started with the defaults. It usually works great, but I'm processing many pdfs, and there are cases where wonky things happen. In file I provided, I get the following table:
<img width="555" alt="image" src="https://user-images.githubusercontent.com/17770722/225011094-751432a7-a10b-47ac-b77a-dc7fb3e71cfb.png">

Notice how some of the columns are shifted over? (like Total Aggregates)
I have done extensive debugging inside of pdfplumber, and this appears to be caused by extra rectangles. Compare "Monomer Purity" to "Total Aggregates" in your image, and you can see the extra rectangles on the left. 

This is just one example of many. Really, I blame the pdf creation for being weird, but I have no control over that.  However, by following the actual visible lines (which are rects instead of lines), I get perfect output.

**Reply by jsvine** [reply]
Yeah, that's a funky PDF! Lots of invisible, confounding rects — `im.reset().draw_rects(page.rects)`:

![image](https://user-images.githubusercontent.com/534702/225052294-7f254792-8228-4250-8f2a-defe1a059c1a.png)


In this case, I'd suggest something like the following:

```python
def reject_2d_rects(obj):
    is_rect = obj["object_type"] == "rect"
    is_thin = obj["width"] < 1 or obj["height"] < 1
    return not (is_rect and not is_thin)

filtered = page.filter(reject_2d_rects)
filtered.to_image().draw_rects(filtered.rects)
```

![image](https://user-images.githubusercontent.com/534702/225052390-891f76ba-bfda-4708-aaa2-237483c38bc5.png)

... and then running `filtered.extract_tables()` gets you what I *think* you want?:

```
[[['Analytical Property', 'Unit', 'Storage\nCondition', 'Time Point'],
  ['Quantity Tests', None, None, None],
  ['Quantity (UV/VIS)', '% Label Claim', '', ''],
  [None, 'mg/mL', '', ''],
  ['Potency Tests', None, None, None],
  ['Potency (Bioassay)', '%', '', ''],
  ['Purity Tests', None, None, None],
  ['Monomer Purity (SEC)', '%', '', ''],
  ['Total Aggregates (SEC)', '%', '', ''],
  ['Reduced Purity\n(Reduced CE-SDS)', '%', '', ''],
  ['Non-Reduced Purity\n(Non-Reduced CE-SDS)', '%', '', ''],
  ['Other Tests', None, None, None],
  ['Description (Visual)', 'N/A', '', ''],
  ['Color', 'N/A', '', ''],
  ['Clarity', 'NTU', '', '']]]
```

**Reply by bellma-lilly** [reply]
@jsvine I think you are right that this approach probably gets to the same place. Intuitively, I like the idea of using the visual lines. No problem if you don't want to go that route. Thanks for providing an alternative that doesn't require altering pdfplumber code.

**Reply by jsvine** [reply]
Great! For context: My general approach with the table extraction is to make as few assumptions as possible about internal structure. In the PDF world, it seems that every improvement for handling one set of PDFs ends up creating new, troublesome edge-cases for others 🫠
